src/services/interfaces/aof.rs

```diff
@@ -0,0 +1,130 @@
+use anyhow::Result;
+use bytes::{BufMut, Bytes, BytesMut};
+
+/// Trait for an Append-Only File (AOF) abstraction.
+
+pub trait TAof {
+    /// Appends a single `AofEntry` to the log.
+    fn append(
+        &mut self,
+        op: &WriteOperation,
+    ) -> impl std::future::Future<Output = Result<()>> + Send;
+
+    /// Replays all logged entries from the beginning of the AOF, calling the provided callback `f`
+    /// for each entry.
+    ///
+    /// The callback `f(AofEntry)` receives each entry in the order it was appended.
+    fn replay<F>(&mut self, f: F) -> impl std::future::Future<Output = Result<()>> + Send
+    where
+        F: FnMut(AofEntry) + Send;
+
+    /// Forces pending writes to be physically recorded on disk.
+    fn fsync(&mut self) -> impl std::future::Future<Output = Result<()>> + Send;
+
+    /// Returns the highest offset **successfully** appended or replayed so far.
+    ///
+    /// - If zero, means no entries were successfully written.
+    /// - This offset should never skip a number if the implementation fails (e.g., fails mid-write).
+    fn latest_offset(&self) -> u64;
+}
+
+/// A single entry in the Append-Only File (AOF).
+#[derive(Debug, Clone, PartialEq, Eq)]
```

Comment:
if we remove this thin wrapper, `AofEntry` and have `WriteOperation`, how would this be different from `QueryIO`?

Any expected, additional role for this?

Reply:
I don't think there is a huge difference now, but I'd prefer having a separate type that represents AOF-specific operations (though perhaps share some implementation?). Curious if you have strong feelings on using `QueryIO` here.

Reply:
@dpjungmin the way we serialize/ deserialize data is implementation detail - which is also domain logic. 

One line of change and when it is breaking change, it should break everything that depends on that logic - not partial failure. 

so, it doesn't have to be QueryIO or AOF, but that serialization and deserialization logic should be in one place. How do you think?

Reply:
Yup, 100% agree with that. So, we would want to keep any serialize/deserialize implementation within services rather than within individual adapters. Is that a correct read back?

Reply:
@dpjungmin right!